pandas/index.pyx

```diff
@@ -267,29 +263,37 @@ cdef class IndexEngine:
         hash(val)
 
     cdef inline _ensure_mapping_populated(self):
-        # need to reset if we have previously
-        # set the initialized from monotonic checks
-        if self.unique_check:
-            self.initialized = 0
-        if not self.initialized:
+        # populate the mappings
+
+        if self.need_unique_check or not self.initialized:
```

Comment:
Do we still need to check against `need_unique_check` here, or is `self.initialized` sufficient now?